src/runtime/cpu/kernels/kernels.cpp

```diff
@@ -78,3 +78,56 @@ float deepworks::CPUNLLLoss(Matrix predictions, ConstVector target) {
     float loss = -X_1d(0, slice).array().sum() / static_cast<float>(batch_size);
     return loss;
 }
+
+void deepworks::CPUBatchNorm1DForward(ConstMatrix input, Matrix output,
```

Comment:
Please separate computing gradients by inputs from parameters, because if layer is frozen it isn't necessary to compute parameters gradients but computing gradient by inputs is mandatory and vice versa if the layer is input no need to compute gradient by inputs, but if isn't frozen need to compute by parameters.

Have a look on PR with `backward`